text/0000-project-ffi-unwind.md

```diff
@@ -0,0 +1,211 @@
+possibility. The problem is that there is no alternative available to them that
+would allow them to keep working (even if they are continuing to rely on
+behavior that is not yet fully specified).
+
+Previous attempts to provide a well-defined mechanism for unwinding across FFI
+boundaries have failed to reach consensus. Notably, two proposed RFCs generated
+over 400 comments between them before ultimately being closed:
+
+* [rust-lang/rfcs#2699](https://github.com/rust-lang/rfcs/pull/2699)
+* [rust-lang/rfcs#2753](https://github.com/rust-lang/rfcs/pull/2753)
+
+GitHub comment threads become difficult to follow for discussions this lengthy,
+and the disagreements in these threads have felt less productive than we
+believe they could be if more structure were provided.
+
+We would also like to demonstrate the Rust lang team's commitment to providing
+such a mechanism without needing to agree in advance on what language changes
+will be stabilized in order to do so.
+
+# Prototyping 'shepherded' project groups
+[prototyping-project-groups]: #prototyping-shepherded-project-groups
+
+With this RFC, we formally announce the formation of a project-specific,
+shepherded "project group" to adopt responsibility for driving progress on
+specifying unwinding behavior at FFI boundaries.
+
+## What is a "project group"?
+
+The "project group" term has not previously been used: it is intended to
+formalize a concept that has existed informally for some time, under a number
+of names (including "working group").
+
+A "project group" is a group of people working on a particular project at the
+behest of an official Rust team. Project groups must have:
+
+* A **charter** defining the project's scope
+* A **liaison** with an official Rust team (who may or may not also be a shepherd)
+* A small number of **shepherds**, who are responsible for summarizing
+  conversations and keeping the lang team abreast of interesting developments.
+* A GitHub repository hosted under the `rust-lang` organization containing the
+  charter and instructions for how community members can monitor the group's
+  progress and/or participate.
+
+[This blog post][shepherds-3.0] explains in detail the role of the
+shepherds.
+
+## Project group roadmap and RFCs
+
+The first step of the project group is to define a **roadmap** indicating the
+planned sequence in which it will design and propose particular behaviors and
+features.  Once the project group feels it has completed work on some item in
+the roadmap, that item will be submitted to the community as an RFC or FCP.
+
+## Stabilizing unspecified "TBD" behavior
+[stabilizing-tbd]: stabilizing-unspecified-tbd-behavior
+
+We would like to be able to provide features in stable Rust without a full
+formal specification but with an informal statement of intent regarding the
+behavior. These features would be considered "unspecified behavior" (rather
+than "undefined behavior"), and their behavior would therefore be subject to
```

Comment:
As worded right now, this sort of assumes that everyone agrees what "unspecified" means (that often turns out not to be the case). I assume this will be detailed and clarified as the project progresses. I hope this will not turn into "implementation defined" or some such.

Reply:
We have indeed written up [our own definition](https://github.com/rust-lang/project-ffi-unwind/blob/master/spec-terminology.md), and I think it would be worthwhile to open an RFC to concretely define this terminology.

I think "implementation defined" would be a good addition to the `spec-terminology` doc.

Reply:
An alternative is for "TBD" to be an orthogonal denotation to whatever the behavior class used by the reference is (unspecified, undefined, target-dependent, ...). 

One of the consequences of unspecified meaning different things to different people (and different WGs, the UCGs has its own definitions and they are controversial!), is that users might end up confused on whether "unspecified" behavior can become "undefined" in the future or not.

I don't think that's something that's worth resolving right now, and I think that it is in the best interest of the Project Group to have as much freedom as possible. By making it an orthogonal annotation, the Project Group can say that "X is UB (TBD) and Y is target-dependent (TBD)" meaning that X is UB and Y is target dependent, and that the project group is trying to improve both, but while X might remain UB, Y won't become worse than target-dependent.

Reply:
Hm. I appreciate the suggestion, @gnzlbg, but I don't really think that helps with the problem Niko and I were trying to solve by defining/adopting "TBD", which is to have a way of saying, essentially, "this is _not_ UB in the sense that the compiler may optimize based on the assumption that it can't happen; we have not fully specified the behavior, but you can expect it to 'work' approximately as you'd expect, and we promise to further specify the behavior later (though until that time the details of the behavior-as-implemented may change)."

I don't think it would be appropriate to define "unspecified behavior" as part of this "announcement" RFC, but if @Centril (or someone else) would prefer, I can just remove this section entirely (assuming @nikomatsakis agrees that's okay).

Reply:
That would mean that any behavior that we might want to define in some way, and one of the alternatives being considered is defining it as "UB", would need to be UB instead of TBD. If that's the intent, then that's fine with me. Maybe add a sentence mentioning just that ?

Reply:
@BatmanAoD I'm just happy with the knowledge that you'll take into account that I don't think this should become implementation-defined and that the lang team should have the freedom to change the Rust unwind ABI on any platform including with shims if necessary.

Reply:
My two cents:

I think it is useful to enumerate our scope and goals and to communicate those goals to people. So I think that distinguishing TBD is very useful, as it communicates that. I think that if we can give people guidelines as to our *intent* (along the lines that @BatmanAoD sketched -- perhaps more specific) that's even better.

That said, I think we should leave ourselves the freedom to make decisions. I think we can define our intent here as roughly:

* We intend to make it possible for Rust panics and foreign exceptions to propagate through a "C unwind" boundary in a well-defined and specified way.
* As a non-exhaustive list, this means that we intend to eventually use cases like the following:
    * a C++ exception unwinding Rust frames, running destructors as it goes
    * a Rust panic unwinding through C++ frames, running destructors as it goes
    * a Rust panic unwinding into C++ and then back into Rust, and being caught via `catch_unwind`
* However, there are a number of details whose resolution is not clear. As a non-exhaustive list:
    * what should happen when a C++ exception attempts to unwind through a Rust `catch_unwind` call?
    * how should longjmp, which is defined on some platforms via unwinding, interact with frames that contain destructors? (And is that even in scope for this group?)

Along those lines, I think we can specify that:

* For Tier 1 platforms, there will be *some* translation from Rust panics into the native unwinding mechanism, but we have not yet specified how Rust panics will be described
* Similarly, we have not specified what the behavior will be when unwinding across a "C unwind" barrier occurs, except for the case where the exception was originally a Rust panic, in which case it is translated back into a Rust panic
    * This implies that we have not specified what happens when `catch_unwind` is involved, since before `catch_unwind` gets involved, there must be a "C unwind" barrier.

I'm going to be honest and say that the distinction between "unspecified behavior" and "undefined behavior" feels like a "distinction without a difference" to me, most of the time, unless that unspecified behavior is further qualified to some narrower range of possibilities. I guess I would mostly rather we just don't waste a lot of time (too late...) arguing back and forth on this, as I'm not sure what practical import it has. I'd rather we just spend our time deciding what the behavior *ought to be*.